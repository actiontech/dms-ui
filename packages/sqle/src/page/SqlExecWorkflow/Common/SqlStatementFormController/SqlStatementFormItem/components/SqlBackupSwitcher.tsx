import {
  FormItemLabel,
  CustomLabelContent
} from '@actiontech/shared/lib/components/FormCom';
import { useTranslation } from 'react-i18next';
import { EmptyBox, BasicInputNumber, BasicToolTips } from '@actiontech/shared';
import { Form } from 'antd';
import { SqlAuditInfoFormFields } from '../../../../Create/index.type';
import { SqlBackupSwitcherProps } from './index.type';
import { CreateAuditTasksGroupReqV1ExecModeEnum } from '@actiontech/shared/lib/api/sqle/service/common.enum';
import { useEffect, useMemo, useCallback } from 'react';
import useCreationMode from '../../../../Create/hooks/useCreationMode';
import SwitchField from './SwitchField';
import { InfoCircleOutlined } from '@actiontech/icons';
import { sortBy } from 'lodash';

const SqlBackupSwitcher: React.FC<SqlBackupSwitcherProps> = ({
  fieldPrefixPath,
  databaseInfo,
  isSameSqlForAll,
  isAtRejectStep,
  isAtFormStep,
  isAuditing
}) => {
  const { t } = useTranslation();

  const form = Form.useFormInstance<SqlAuditInfoFormFields>();

  const { isCloneMode } = useCreationMode();

  const currentExecuteMode = Form.useWatch(
    [fieldPrefixPath, 'exec_mode'],
    form
  );

  const enableBackup = Form.useWatch([fieldPrefixPath, 'backup'], form);

  const getInstanceEnableBackup = useCallback(() => {
    if (isSameSqlForAll) {
      return databaseInfo.some((item) => item.enableBackup);
    }
    return (
      databaseInfo.find((item) => item.key === fieldPrefixPath)?.enableBackup ??
      false
    );
  }, [databaseInfo, fieldPrefixPath, isSameSqlForAll]);

  const getInstanceBackupMaxRows = useCallback(() => {
    if (isSameSqlForAll) {
      const enableBackupInstances = databaseInfo.filter((i) => i.enableBackup);
      // 相同SQL模式下 多个数据源的行数限制选最小值
      return !!enableBackupInstances.length
        ? sortBy(enableBackupInstances, 'backupMaxRows')?.[0]?.backupMaxRows
        : 1000;
    }
    return (
      databaseInfo.find((item) => item.key === fieldPrefixPath)
        ?.backupMaxRows ?? 1000
    );
  }, [databaseInfo, fieldPrefixPath, isSameSqlForAll]);

  const enableBackupInstanceName = useMemo(() => {
    return databaseInfo
      .filter((i) => i.enableBackup)
      .map((i) => i.instanceName)
      .join(',');
  }, [databaseInfo]);

  useEffect(() => {
    if (isAtFormStep && !isAuditing.value && !isCloneMode) {
      const currentEnableBackup = getInstanceEnableBackup();
      if (!form.isFieldTouched([fieldPrefixPath, 'backup'])) {
        form.setFieldValue([fieldPrefixPath, 'backup'], currentEnableBackup);
      }

      if (!form.isFieldTouched([fieldPrefixPath, 'backupMaxRows'])) {
        form.setFieldValue(
          [fieldPrefixPath, 'backupMaxRows'],
          getInstanceBackupMaxRows()
        );
      }
    }
  }, [
    fieldPrefixPath,
    getInstanceEnableBackup,
    form,
    isCloneMode,
    isAtFormStep,
    isAuditing,
    getInstanceBackupMaxRows
  ]);

  const allowBackup = useMemo(() => {
    return databaseInfo.find((item) => item.key === fieldPrefixPath)
      ?.allowBackup;
  }, [databaseInfo, fieldPrefixPath]);

  return (
    <EmptyBox
      if={
        currentExecuteMode !==
          CreateAuditTasksGroupReqV1ExecModeEnum.sql_file && allowBackup
      }
    >
      <FormItemLabel
        className="has-label-tip"
        label={
          <CustomLabelContent
            title={t('execWorkflow.create.form.sqlInfo.switchSqlBackup')}
            tips={t('execWorkflow.create.form.sqlInfo.switchSqlBackupTips')}
          />
        }
        labelCol={{ span: 22 }}
        wrapperCol={{ span: 2 }}
        name={[fieldPrefixPath, 'backup']}
        valuePropName="checked"
      >
        <SwitchField
          title={
            isSameSqlForAll
              ? t(
                  'execWorkflow.create.form.sqlInfo.cancelSwitchSqlBackupTipsWithInstanceName',
                  {
                    instanceName: enableBackupInstanceName
                  }
                )
              : t('execWorkflow.create.form.sqlInfo.cancelSwitchSqlBackupTips')
          }
          disabled={isAtRejectStep}
        />
      </FormItemLabel>
      <EmptyBox if={enableBackup}>
        <FormItemLabel
          className="has-label-tip"
          label={
            <CustomLabelContent
              title={
                <BasicToolTips
                  title={t(
                    'execWorkflow.create.form.sqlInfo.backupMaxRowsLimitTooltips'
                  )}
                  suffixIcon={<InfoCircleOutlined width={14} height={14} />}
                >
                  {t('execWorkflow.create.form.sqlInfo.backupMaxRowsLimit')}
                </BasicToolTips>
              }
              tips={t(
                'execWorkflow.create.form.sqlInfo.backupMaxRowsLimitTips'
              )}
            />
          }
          labelCol={{ span: 14 }}
          wrapperCol={{ span: 10 }}
          name={[fieldPrefixPath, 'backupMaxRows']}
          initialValue={1000}
        >
          <BasicInputNumber min={0} disabled={getInstanceEnableBackup()} />
        </FormItemLabel>
      </EmptyBox>
    </EmptyBox>
  );
};

export default SqlBackupSwitcher;
