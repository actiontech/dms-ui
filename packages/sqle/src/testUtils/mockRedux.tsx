import whitelist from '../store/whitelist';
import React from 'react';
import { Provider } from 'react-redux';
import { combineReducers, configureStore } from '@reduxjs/toolkit';
import { Dictionary } from '@actiontech/shared/lib/types/common.type';
import sqlManagement from '../store/sqlManagement';
import sqlManagementException from '../store/sqlManagementException';
import pipeline from '../store/pipeline';

const reducers = combineReducers({
  whitelist,
  sqlManagement,
  sqlManagementException,
  pipeline
});

export const storeFactory = (initStore: Dictionary = {}) => {
  return configureStore({
    reducer: reducers,
    preloadedState: initStore
  });
};

export const CustomProvider: React.FC<{
  initStore: Dictionary;
  children: JSX.Element;
}> = (props) => {
  return (
    <Provider store={storeFactory(props.initStore)}>{props.children}</Provider>
  );
};
