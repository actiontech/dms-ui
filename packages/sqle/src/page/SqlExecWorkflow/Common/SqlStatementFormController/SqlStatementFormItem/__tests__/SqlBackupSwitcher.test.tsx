import SqlBackupSwitcher from '../components/SqlBackupSwitcher';
import { fireEvent, renderHook, screen, act } from '@testing-library/react';
import { Form } from 'antd';
import { SqlBackupSwitcherProps } from '../components/index.type';
import { sqleSuperRender } from '../../../../../../testUtils/superRender';
import { getBySelector } from '@actiontech/shared/lib/testUtil/customQuery';

describe('test SqlBackupSwitcher', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useFakeTimers();
  });

  const mockDatabaseInfo = [
    {
      key: '1',
      instanceName: 'mysql-1',
      schemaName: 'test',
      enableBackup: true,
      allowBackup: true,
      backupMaxRows: 2000
    },
    {
      key: '2',
      instanceName: 'mysql-2',
      schemaName: 'test2',
      enableBackup: false,
      allowBackup: true,
      backupMaxRows: 3000
    }
  ];

  const customRender = (
    databaseInfo: SqlBackupSwitcherProps['databaseInfo'],
    isSameSqlForAll = false,
    isAtRejectStep = false,
    isAuditing = false,
    isAtFormStep = true,
    search = ''
  ) => {
    const { result } = renderHook(() => Form.useForm());
    return sqleSuperRender(
      <Form form={result.current[0]}>
        <SqlBackupSwitcher
          fieldPrefixPath="1"
          databaseInfo={databaseInfo}
          isSameSqlForAll={isSameSqlForAll}
          isAtRejectStep={isAtRejectStep}
          isAuditing={{ set: jest.fn(), value: isAuditing }}
          isAtFormStep={isAtFormStep}
        />
      </Form>,
      undefined,
      {
        routerProps: {
          initialEntries: [
            {
              search
            }
          ]
        }
      }
    );
  };

  it('render init snap', () => {
    const { baseElement } = customRender([]);
    expect(baseElement).toMatchSnapshot();
    expect(screen.queryByText('是否选择开启备份')).not.toBeInTheDocument();
  });

  it('render switch when isAtRejectStep is true', () => {
    const { baseElement } = customRender(mockDatabaseInfo, false, true);
    expect(baseElement).toMatchSnapshot();
    expect(screen.getByRole('switch')).toHaveAttribute('disabled');
  });

  it('render default checked when isSameSqlForAll is true', async () => {
    const { baseElement } = customRender(mockDatabaseInfo, true);
    expect(screen.getByRole('switch').getAttribute('aria-checked')).toBe(
      'true'
    );
    expect(screen.getByText('回滚行数限制')).toBeInTheDocument();
    expect(getBySelector('.ant-input-number-input')).toBeDisabled();
    expect(getBySelector('.ant-input-number-input')).toHaveValue('2000');
    expect(baseElement).toMatchSnapshot();
    fireEvent.click(getBySelector('button'));
    await act(async () => jest.advanceTimersByTime(300));
    expect(screen.getByText('mysql-1已开启备份需求，是否确认关闭备份？'));
    fireEvent.click(screen.getByText('确 认'));
    expect(screen.getByRole('switch').getAttribute('aria-checked')).toBe(
      'false'
    );
    expect(screen.queryByText('回滚行数限制')).not.toBeInTheDocument();
  });

  it('render default checked when isAtFormStep is false', async () => {
    customRender(mockDatabaseInfo, true, false, false, false);
    expect(screen.getByRole('switch').getAttribute('aria-checked')).toBe(
      'false'
    );
  });

  it('render default checked when isSameSqlForAll is true and isAuditing is true', async () => {
    customRender(mockDatabaseInfo, true, false, true);
    expect(screen.getByRole('switch').getAttribute('aria-checked')).toBe(
      'false'
    );
  });

  it('render default checked when isSameSqlForAll is true and isCloneMode is true', async () => {
    customRender(
      mockDatabaseInfo,
      true,
      false,
      false,
      true,
      '?sourceWorkflowId=111'
    );
    expect(screen.getByRole('switch').getAttribute('aria-checked')).toBe(
      'false'
    );
  });

  it('render default checked when isSameSqlForAll is false', async () => {
    const { baseElement } = customRender(mockDatabaseInfo);
    expect(screen.getByRole('switch').getAttribute('aria-checked')).toBe(
      'true'
    );
    expect(baseElement).toMatchSnapshot();
    fireEvent.click(getBySelector('button'));
    await act(async () => jest.advanceTimersByTime(300));
    expect(
      screen.getByText('当前应用的数据源已开启备份需求，是否确认关闭备份？')
    );
    fireEvent.click(screen.getByText('确 认'));
    expect(screen.getByRole('switch').getAttribute('aria-checked')).toBe(
      'false'
    );
  });

  it('render snap when database is not allow backup', async () => {
    const { baseElement } = customRender(
      [
        {
          key: '1',
          instanceName: 'mysql-1',
          schemaName: 'test',
          enableBackup: false,
          allowBackup: false
        },
        {
          key: '2',
          instanceName: 'mysql-2',
          schemaName: 'test2',
          enableBackup: false,
          allowBackup: false
        }
      ],
      true
    );
    expect(screen.queryByText('是否选择开启备份')).not.toBeInTheDocument();
    expect(screen.queryByText('回滚行数限制')).not.toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  it('render snap when database allow backup but not enable backup', async () => {
    const { baseElement } = customRender(
      [
        {
          key: '1',
          instanceName: 'mysql-1',
          schemaName: 'test',
          enableBackup: false,
          allowBackup: true
        },
        {
          key: '2',
          instanceName: 'mysql-2',
          schemaName: 'test2',
          enableBackup: false,
          allowBackup: true
        }
      ],
      true
    );
    expect(screen.getByText('是否选择开启备份')).toBeInTheDocument();
    expect(screen.getByRole('switch').getAttribute('aria-checked')).toBe(
      'false'
    );
    expect(screen.queryByText('回滚行数限制')).not.toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
    fireEvent.click(getBySelector('button'));
    await act(async () => jest.advanceTimersByTime(300));
    expect(screen.getByRole('switch').getAttribute('aria-checked')).toBe(
      'true'
    );
    expect(screen.getByText('回滚行数限制')).toBeInTheDocument();
    expect(getBySelector('.ant-input-number-input')).not.toBeDisabled();
    expect(getBySelector('.ant-input-number-input')).toHaveValue('1000');
  });
});
