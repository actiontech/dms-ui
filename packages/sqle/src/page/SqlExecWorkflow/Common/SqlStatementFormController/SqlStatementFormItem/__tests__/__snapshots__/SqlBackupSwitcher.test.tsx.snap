// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test SqlBackupSwitcher render default checked when isSameSqlForAll is false 1`] = `
<body>
  <div />
  <div>
    <form
      class="ant-form ant-form-horizontal css-dev-only-do-not-override-txh9fw"
    >
      <div
        class="ant-form-item has-label-tip css-1jlm9cy css-dev-only-do-not-override-txh9fw"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-txh9fw"
        >
          <div
            class="ant-col ant-col-22 ant-form-item-label css-dev-only-do-not-override-txh9fw"
          >
            <label
              class=""
              for="1_backup"
              title=""
            >
              <div
                class="label-cont-custom"
              >
                <div>
                  是否选择开启备份
                </div>
                <div
                  class="tip-content-box"
                >
                  开启后，数据源上创建的工单将默认开启备份能力
                </div>
              </div>
            </label>
          </div>
          <div
            class="ant-col ant-col-2 ant-form-item-control css-dev-only-do-not-override-txh9fw"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  aria-checked="true"
                  class="ant-switch basic-switch-wrapper backup-switcher css-g6dhbn css-dev-only-do-not-override-txh9fw ant-switch-checked"
                  role="switch"
                  type="button"
                >
                  <div
                    class="ant-switch-handle"
                  />
                  <span
                    class="ant-switch-inner"
                  >
                    <span
                      class="ant-switch-inner-checked"
                    />
                    <span
                      class="ant-switch-inner-unchecked"
                    />
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item has-label-tip css-1jlm9cy css-dev-only-do-not-override-txh9fw"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-txh9fw"
        >
          <div
            class="ant-col ant-col-14 ant-form-item-label css-dev-only-do-not-override-txh9fw"
          >
            <label
              class=""
              for="1_backupMaxRows"
              title=""
            >
              <div
                class="label-cont-custom"
              >
                <div>
                  回滚行数限制
                </div>
                <div
                  class="tip-content-box"
                >
                  <div
                    class="ant-space css-dev-only-do-not-override-txh9fw ant-space-horizontal ant-space-align-center"
                  >
                    <div
                      class="ant-space-item"
                    >
                      <div
                        class="ant-space css-dev-only-do-not-override-txh9fw ant-space-horizontal ant-space-align-center basic-tooltips-wrapper css-4jbtl9"
                      >
                        <div
                          class="ant-space-item"
                          style="margin-right: 8px;"
                        >
                          当预计影响行数超过指定值则不回滚
                        </div>
                        <div
                          class="ant-space-item"
                        >
                          <svg
                            height="14"
                            viewBox="0 0 14 14"
                            width="14"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M7 12.833A5.833 5.833 0 1 1 7 1.167a5.833 5.833 0 0 1 0 11.666m0-1.166a4.666 4.666 0 1 0 0-9.333 4.666 4.666 0 0 0 0 9.333m-.583-7.584h1.166V5.25H6.417zm0 2.334h1.166v3.5H6.417z"
                              fill="#EDB054"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </label>
          </div>
          <div
            class="ant-col ant-col-10 ant-form-item-control css-dev-only-do-not-override-txh9fw"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-input-number ant-input-number-lg ant-input-number-in-form-item css-dev-only-do-not-override-ezmdel basic-inputNumber-wrapper css-1fd5hnt ant-input-number-disabled"
                >
                  <div
                    class="ant-input-number-handler-wrap"
                  >
                    <span
                      aria-disabled="false"
                      aria-label="Increase Value"
                      class="ant-input-number-handler ant-input-number-handler-up"
                      role="button"
                      unselectable="on"
                    >
                      <span
                        aria-label="up"
                        class="anticon anticon-up ant-input-number-handler-up-inner"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="up"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                    <span
                      aria-disabled="false"
                      aria-label="Decrease Value"
                      class="ant-input-number-handler ant-input-number-handler-down"
                      role="button"
                      unselectable="on"
                    >
                      <span
                        aria-label="down"
                        class="anticon anticon-down ant-input-number-handler-down-inner"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="down"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                  </div>
                  <div
                    class="ant-input-number-input-wrap"
                  >
                    <input
                      aria-valuemin="0"
                      aria-valuenow="2000"
                      autocomplete="off"
                      class="ant-input-number-input"
                      disabled=""
                      id="1_backupMaxRows"
                      role="spinbutton"
                      step="1"
                      value="2000"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</body>
`;

exports[`test SqlBackupSwitcher render default checked when isSameSqlForAll is true 1`] = `
<body>
  <div />
  <div>
    <form
      class="ant-form ant-form-horizontal css-dev-only-do-not-override-txh9fw"
    >
      <div
        class="ant-form-item has-label-tip css-1jlm9cy css-dev-only-do-not-override-txh9fw"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-txh9fw"
        >
          <div
            class="ant-col ant-col-22 ant-form-item-label css-dev-only-do-not-override-txh9fw"
          >
            <label
              class=""
              for="1_backup"
              title=""
            >
              <div
                class="label-cont-custom"
              >
                <div>
                  是否选择开启备份
                </div>
                <div
                  class="tip-content-box"
                >
                  开启后，数据源上创建的工单将默认开启备份能力
                </div>
              </div>
            </label>
          </div>
          <div
            class="ant-col ant-col-2 ant-form-item-control css-dev-only-do-not-override-txh9fw"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  aria-checked="true"
                  class="ant-switch basic-switch-wrapper backup-switcher css-g6dhbn css-dev-only-do-not-override-txh9fw ant-switch-checked"
                  role="switch"
                  type="button"
                >
                  <div
                    class="ant-switch-handle"
                  />
                  <span
                    class="ant-switch-inner"
                  >
                    <span
                      class="ant-switch-inner-checked"
                    />
                    <span
                      class="ant-switch-inner-unchecked"
                    />
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item has-label-tip css-1jlm9cy css-dev-only-do-not-override-txh9fw"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-txh9fw"
        >
          <div
            class="ant-col ant-col-14 ant-form-item-label css-dev-only-do-not-override-txh9fw"
          >
            <label
              class=""
              for="1_backupMaxRows"
              title=""
            >
              <div
                class="label-cont-custom"
              >
                <div>
                  回滚行数限制
                </div>
                <div
                  class="tip-content-box"
                >
                  <div
                    class="ant-space css-dev-only-do-not-override-txh9fw ant-space-horizontal ant-space-align-center"
                  >
                    <div
                      class="ant-space-item"
                    >
                      <div
                        class="ant-space css-dev-only-do-not-override-txh9fw ant-space-horizontal ant-space-align-center basic-tooltips-wrapper css-4jbtl9"
                      >
                        <div
                          class="ant-space-item"
                          style="margin-right: 8px;"
                        >
                          当预计影响行数超过指定值则不回滚
                        </div>
                        <div
                          class="ant-space-item"
                        >
                          <svg
                            height="14"
                            viewBox="0 0 14 14"
                            width="14"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M7 12.833A5.833 5.833 0 1 1 7 1.167a5.833 5.833 0 0 1 0 11.666m0-1.166a4.666 4.666 0 1 0 0-9.333 4.666 4.666 0 0 0 0 9.333m-.583-7.584h1.166V5.25H6.417zm0 2.334h1.166v3.5H6.417z"
                              fill="#EDB054"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </label>
          </div>
          <div
            class="ant-col ant-col-10 ant-form-item-control css-dev-only-do-not-override-txh9fw"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-input-number ant-input-number-lg ant-input-number-in-form-item css-dev-only-do-not-override-ezmdel basic-inputNumber-wrapper css-1fd5hnt ant-input-number-disabled"
                >
                  <div
                    class="ant-input-number-handler-wrap"
                  >
                    <span
                      aria-disabled="false"
                      aria-label="Increase Value"
                      class="ant-input-number-handler ant-input-number-handler-up"
                      role="button"
                      unselectable="on"
                    >
                      <span
                        aria-label="up"
                        class="anticon anticon-up ant-input-number-handler-up-inner"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="up"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                    <span
                      aria-disabled="false"
                      aria-label="Decrease Value"
                      class="ant-input-number-handler ant-input-number-handler-down"
                      role="button"
                      unselectable="on"
                    >
                      <span
                        aria-label="down"
                        class="anticon anticon-down ant-input-number-handler-down-inner"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="down"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                  </div>
                  <div
                    class="ant-input-number-input-wrap"
                  >
                    <input
                      aria-valuemin="0"
                      aria-valuenow="2000"
                      autocomplete="off"
                      class="ant-input-number-input"
                      disabled=""
                      id="1_backupMaxRows"
                      role="spinbutton"
                      step="1"
                      value="2000"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</body>
`;

exports[`test SqlBackupSwitcher render init snap 1`] = `
<body>
  <div />
  <div>
    <form
      class="ant-form ant-form-horizontal css-dev-only-do-not-override-txh9fw"
    />
  </div>
</body>
`;

exports[`test SqlBackupSwitcher render snap when database allow backup but not enable backup 1`] = `
<body>
  <div />
  <div>
    <form
      class="ant-form ant-form-horizontal css-dev-only-do-not-override-txh9fw"
    >
      <div
        class="ant-form-item has-label-tip css-1jlm9cy css-dev-only-do-not-override-txh9fw"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-txh9fw"
        >
          <div
            class="ant-col ant-col-22 ant-form-item-label css-dev-only-do-not-override-txh9fw"
          >
            <label
              class=""
              for="1_backup"
              title=""
            >
              <div
                class="label-cont-custom"
              >
                <div>
                  是否选择开启备份
                </div>
                <div
                  class="tip-content-box"
                >
                  开启后，数据源上创建的工单将默认开启备份能力
                </div>
              </div>
            </label>
          </div>
          <div
            class="ant-col ant-col-2 ant-form-item-control css-dev-only-do-not-override-txh9fw"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  aria-checked="false"
                  class="ant-switch basic-switch-wrapper backup-switcher css-g6dhbn css-dev-only-do-not-override-txh9fw"
                  role="switch"
                  type="button"
                >
                  <div
                    class="ant-switch-handle"
                  />
                  <span
                    class="ant-switch-inner"
                  >
                    <span
                      class="ant-switch-inner-checked"
                    />
                    <span
                      class="ant-switch-inner-unchecked"
                    />
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</body>
`;

exports[`test SqlBackupSwitcher render snap when database is not allow backup 1`] = `
<body>
  <div />
  <div>
    <form
      class="ant-form ant-form-horizontal css-dev-only-do-not-override-txh9fw"
    />
  </div>
</body>
`;

exports[`test SqlBackupSwitcher render switch when isAtRejectStep is true 1`] = `
<body>
  <div />
  <div>
    <form
      class="ant-form ant-form-horizontal css-dev-only-do-not-override-txh9fw"
    >
      <div
        class="ant-form-item has-label-tip css-1jlm9cy css-dev-only-do-not-override-txh9fw"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-txh9fw"
        >
          <div
            class="ant-col ant-col-22 ant-form-item-label css-dev-only-do-not-override-txh9fw"
          >
            <label
              class=""
              for="1_backup"
              title=""
            >
              <div
                class="label-cont-custom"
              >
                <div>
                  是否选择开启备份
                </div>
                <div
                  class="tip-content-box"
                >
                  开启后，数据源上创建的工单将默认开启备份能力
                </div>
              </div>
            </label>
          </div>
          <div
            class="ant-col ant-col-2 ant-form-item-control css-dev-only-do-not-override-txh9fw"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  aria-checked="true"
                  class="ant-switch basic-switch-wrapper backup-switcher css-g6dhbn css-dev-only-do-not-override-txh9fw ant-switch-checked ant-switch-disabled"
                  disabled=""
                  role="switch"
                  type="button"
                >
                  <div
                    class="ant-switch-handle"
                  />
                  <span
                    class="ant-switch-inner"
                  >
                    <span
                      class="ant-switch-inner-checked"
                    />
                    <span
                      class="ant-switch-inner-unchecked"
                    />
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item has-label-tip css-1jlm9cy css-dev-only-do-not-override-txh9fw"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-txh9fw"
        >
          <div
            class="ant-col ant-col-14 ant-form-item-label css-dev-only-do-not-override-txh9fw"
          >
            <label
              class=""
              for="1_backupMaxRows"
              title=""
            >
              <div
                class="label-cont-custom"
              >
                <div>
                  回滚行数限制
                </div>
                <div
                  class="tip-content-box"
                >
                  <div
                    class="ant-space css-dev-only-do-not-override-txh9fw ant-space-horizontal ant-space-align-center"
                  >
                    <div
                      class="ant-space-item"
                    >
                      <div
                        class="ant-space css-dev-only-do-not-override-txh9fw ant-space-horizontal ant-space-align-center basic-tooltips-wrapper css-4jbtl9"
                      >
                        <div
                          class="ant-space-item"
                          style="margin-right: 8px;"
                        >
                          当预计影响行数超过指定值则不回滚
                        </div>
                        <div
                          class="ant-space-item"
                        >
                          <svg
                            height="14"
                            viewBox="0 0 14 14"
                            width="14"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M7 12.833A5.833 5.833 0 1 1 7 1.167a5.833 5.833 0 0 1 0 11.666m0-1.166a4.666 4.666 0 1 0 0-9.333 4.666 4.666 0 0 0 0 9.333m-.583-7.584h1.166V5.25H6.417zm0 2.334h1.166v3.5H6.417z"
                              fill="#EDB054"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </label>
          </div>
          <div
            class="ant-col ant-col-10 ant-form-item-control css-dev-only-do-not-override-txh9fw"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-input-number ant-input-number-lg ant-input-number-in-form-item css-dev-only-do-not-override-ezmdel basic-inputNumber-wrapper css-1fd5hnt ant-input-number-disabled"
                >
                  <div
                    class="ant-input-number-handler-wrap"
                  >
                    <span
                      aria-disabled="false"
                      aria-label="Increase Value"
                      class="ant-input-number-handler ant-input-number-handler-up"
                      role="button"
                      unselectable="on"
                    >
                      <span
                        aria-label="up"
                        class="anticon anticon-up ant-input-number-handler-up-inner"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="up"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                    <span
                      aria-disabled="false"
                      aria-label="Decrease Value"
                      class="ant-input-number-handler ant-input-number-handler-down"
                      role="button"
                      unselectable="on"
                    >
                      <span
                        aria-label="down"
                        class="anticon anticon-down ant-input-number-handler-down-inner"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="down"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                  </div>
                  <div
                    class="ant-input-number-input-wrap"
                  >
                    <input
                      aria-valuemin="0"
                      aria-valuenow="2000"
                      autocomplete="off"
                      class="ant-input-number-input"
                      disabled=""
                      id="1_backupMaxRows"
                      role="spinbutton"
                      step="1"
                      value="2000"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</body>
`;
